### PR TITLE
Fix markers

### LINKED ISSUE #108: Date incrementers permit dates outside date min, max
And when they do, erroneous values are shown on the map. This is actually two errors. Fix both.

### LINKED ISSUE #109: Station markers do not reflect actual data
This is a major issue, making this tool incorrect.

To reproduce:

1. Open app.
2. Use the month decrement keys to go back several months.
3. Use them to go forward again.
4. The markers on the map are not the same.

Traced this down to insufficiently unique keys on the circle markers.

### PR DESCRIPTION
**Pull request by rod-glover**
Resolves #108 
Resolves #109

This is an important fix; without it data is displayed inconsistently and wrongly.

Erroneous behaviour was that going back a few months (one at a time, with the -1) control and then moving forward again showed inconsistent (not the same!) markers.

Fix: Now consistent. Reviewer should verify; I can provide support for that.

### CODE REVIEW

**Comment by rod-glover** [comment]
[Demo](http://beehive.pacificclimate.org/weather-anomaly-viewer#)

**Comment by rod-glover** [comment]
Note: I think the upgrade to React 18 may have introduced some different, more "careful" behaviour by React with keys on rendered components. Which meant a shortcoming suddenly became visible a little while ago in my work. This is as well to know for other projects.

**Comment by corviday** [comment]
So the demo and the production version both behave correctly? The error was introduced in a version that has not yet been deployed to production, and rhankfully caught early?

**Comment by rod-glover** [comment]
> So the demo and the production version both behave correctly? The error was introduced in a version that has not yet been deployed to production, and thankfully caught early?

Yes, thank goodness. Earlier, I thought not.

File: src/components/main/Tool.js [hunk 1]
```diff
@@ -29,25 +30,37 @@ import "@/components/main/Tool.css";
...
 
+  const minYear = config.ui.yearSelector.minYear;
+  const minDate = moment({ year: minYear });
+  // There's a tricky and unresolved problem when dates are incremented beyond
+  // lastDateWithMonthlyData: the stations displayed on the map get out of sync
+  // with the actual date and data. This prevents such incrementing, and so the
+  // erroneous maps, but it does not fix the bug causing the error with the station
+  // markers.
+  const maxDate = lastDateWithMonthlyData;
+  const clipDate = (date) => clipMoment(date, minDate, maxDate);
```
Comment (QSparks) [inline src/components/main/Tool.js:58]: I tested incrementing the dates beyond lastDateWithMonthlyData, and I can confirm that this solution works as intended.

File: src/components/map/StationPopup.js [hunk 2]
```diff
@@ -54,7 +55,7 @@ export default function StationPopup({
 
       <div className="border-bottom py-1">
         <div className="fst-italic">
-          <VariableLabel variable={variable} />
+          <VariableLabel variable={variable} /> for {date.format("MMM YYYY")}
```
Comment (QSparks) [inline src/components/map/StationPopup.js:58]: Great addition! I noticed that when I switch to the Baseline display, the year that was set prior to switching is included in the popup.
Possibly from the [type](https://github.com/pacificclimate/weather-anomaly-tool/blob/2387a96aa96e30d805844eb2723dce00a78bdbca/src/components/map/DataMap.js#L127) part of the key?
That said, it's clear from the note: _"Note that baseline values are multi-year (1971-2000) averages for a given period (month)."_

**Review by QSparks — APPROVED** [review]
Great work, Rod, and good catch on discovering this in the first place. It is much cleaner to handle the entire key creation in the uniqueStationKey as you have done. I tested out the demo, flipping between decremented and re-incremented months, and didn’t find any missing stations.

**Comment by corviday** [comment]
Took me a while to understand the actual problem, but here's an example from SE BC.

Start at April 2024:
april1

Click backwards to March 2024:
march1

Click backwards ti February 2024:
feb1

Now click forward to return to March 2024, but notice the colour coding of the stations is different than when we previously viewed March 2024. Perhaps it is retaining the February 2024 coloration?:
mar2

Click forward to return to April 2024, which again does not match the previous viewing of this data:
apr2

I have confirmed this behaviour does not occur in the above-linked demo.

**Comment by rod-glover** [comment]
Excellent, thank you both!

File: src/components/map/StationPopup.js (hunk 2 above)
Comment (rod-glover) [reply src/components/map/StationPopup.js]: No, that's just an oversight. Well spotted. Fixed now.

**Pull request merged**